disk/disk_solaris.go

```diff
@@ -114,7 +116,35 @@ func UsageWithContext(ctx context.Context, path string) (*UsageStat, error) {
 }
 
 func SerialNumberWithContext(ctx context.Context, name string) (string, error) {
-	return "", common.ErrNotImplementedError
+	cfgadm, err := exec.LookPath("cfgadm")
+	if err != nil {
+		return "", fmt.Errorf("find cfgadm: %w", err)
+	}
+	out, err := invoke.CommandWithContext(ctx, cfgadm, "-ls", "select=type(disk),cols=ap_id:info,cols2=,noheadings")
+	if err != nil {
+		return "", fmt.Errorf("exec cfgadm: %w", err)
+	}
+
+	suf := "::" + strings.TrimPrefix(name, "/dev/")
+	s := bufio.NewScanner(bytes.NewReader(out))
+	for s.Scan() {
+		flds := strings.Fields(s.Text())
+		if strings.HasSuffix(flds[0], suf) {
+			flen := len(flds)
+			if flen >= 3 {
+				for i, f := range flds {
+					if i > 0 && i < flen-1 && f == "SN:" {
+						return flds[i+1], nil
+					}
+				}
+			}
+			return "", nil
+		}
+	}
+	if err = s.Err(); err != nil {
```

Comment:
How about this?

```suggestion
	if err := s.Err(); err != nil {
```

Reply:
We already have `err` defined in scope here, and to me reusing it like in the original is cleaner than shadowing it with a new one, so I'd personally keep it as it is in my patch. But your call, do feel free to push the change if you like it better the way you suggested.